BrokenPipeError being occasionally thrown with eventlet worker

In some cases on a very busy application, we get:

```
[2014-11-17 04:18:49 -0500] [13520] [ERROR] Error handling request
Traceback (most recent call last):
  File "/usr/local/lib/python3.4/site-packages/gunicorn/workers/async.py", line 108, in handle_request
    resp.write(item)
  File "/usr/local/lib/python3.4/site-packages/gunicorn/http/wsgi.py", line 344, in write
    util.write(self.sock, arg, self.chunked)
  File "/usr/local/lib/python3.4/site-packages/gunicorn/util.py", line 301, in write
    sock.sendall(data)
  File "/usr/local/lib/python3.4/site-packages/eventlet/greenio.py", line 375, in sendall
    tail = self.send(data, flags)
  File "/usr/local/lib/python3.4/site-packages/eventlet/greenio.py", line 358, in send
    total_sent += fd.send(data[total_sent:], flags)
  File "/usr/local/lib/python3.4/site-packages/eventlet/greenio.py", line 351, in send
    return fd.send(data, flags)
BrokenPipeError: [Errno 32] Broken pipe
```

The reason why is because the user closed out of the tab while it is loading (or otherwise halted navigation).  As such, it is probably not worthwhile to log this as a fatal error.

--- a/gunicorn/workers/async.py
+++ b/gunicorn/workers/async.py
@@ -58,7 +58,7 @@ class AsyncWorker(base.Worker):
                 exc_info = sys.exc_info()
                 # pass to next try-except level
                 six.reraise(exc_info[0], exc_info[1], exc_info[2])
-            except socket.error:
+            except EnvironmentError:
                 exc_info = sys.exc_info()
                 # pass to next try-except level
                 six.reraise(exc_info[0], exc_info[1], exc_info[2])
@@ -71,11 +71,11 @@ class AsyncWorker(base.Worker):
             else:
                 self.log.debug("Error processing SSL request.")
                 self.handle_error(req, client, addr, e)
-        except socket.error as e:
-            if e.args[0] not in (errno.EPIPE, errno.ECONNRESET):
+        except EnvironmentError as e:
+            if e.errno not in (errno.EPIPE, errno.ECONNRESET):
                 self.log.exception("Socket error processing request.")
             else:
-                if e.args[0] == errno.ECONNRESET:
+                if e.errno == errno.ECONNRESET:
                     self.log.debug("Ignoring connection reset")
                 else:
                     self.log.debug("Ignoring EPIPE")
@@ -121,7 +121,7 @@ class AsyncWorker(base.Worker):
                 raise StopIteration()
         except StopIteration:
             raise
-        except socket.error:
+        except EnvironmentError:
             # If the original exception was a socket.error we delegate
             # handling it to the caller (where handle() might ignore it)
             six.reraise(*sys.exc_info())
@@ -133,7 +133,7 @@ class AsyncWorker(base.Worker):
                 try:
                     sock.shutdown(socket.SHUT_RDWR)
                     sock.close()
-                except socket.error:
+                except EnvironmentError:
                     pass
                 raise StopIteration()
             raise

--- a/gunicorn/workers/gthread.py
+++ b/gunicorn/workers/gthread.py
@@ -135,8 +135,8 @@ class ThreadWorker(base.Worker):
             self.nr_conns += 1
             # enqueue the job
             self.enqueue_req(conn)
-        except socket.error as e:
-            if e.args[0] not in (errno.EAGAIN,
+        except EnvironmentError as e:
+            if e.errno not in (errno.EAGAIN,
                     errno.ECONNABORTED, errno.EWOULDBLOCK):
                 raise
 
@@ -176,8 +176,8 @@ class ThreadWorker(base.Worker):
                 with self._lock:
                     try:
                         self.poller.unregister(conn.sock)
-                    except socket.error as e:
-                        if e.args[0] != errno.EBADF:
+                    except EnvironmentError as e:
+                        if e.errno != errno.EBADF:
                             raise
 
                 # close the socket
@@ -287,11 +287,11 @@ class ThreadWorker(base.Worker):
                 self.log.debug("Error processing SSL request.")
                 self.handle_error(req, conn.sock, conn.addr, e)
 
-        except socket.error as e:
-            if e.args[0] not in (errno.EPIPE, errno.ECONNRESET):
+        except EnvironmentError as e:
+            if e.errno not in (errno.EPIPE, errno.ECONNRESET):
                 self.log.exception("Socket error processing request.")
             else:
-                if e.args[0] == errno.ECONNRESET:
+                if e.errno == errno.ECONNRESET:
                     self.log.debug("Ignoring connection reset")
                 else:
                     self.log.debug("Ignoring connection epipe")
@@ -338,7 +338,7 @@ class ThreadWorker(base.Worker):
             if resp.should_close():
                 self.log.debug("Closing connection.")
                 return False
-        except socket.error:
+        except EnvironmentError:
             exc_info = sys.exc_info()
             # pass to next try-except level
             six.reraise(exc_info[0], exc_info[1], exc_info[2])
@@ -350,7 +350,7 @@ class ThreadWorker(base.Worker):
                 try:
                     conn.sock.shutdown(socket.SHUT_RDWR)
                     conn.sock.close()
-                except socket.error:
+                except EnvironmentError:
                     pass
                 raise StopIteration()
             raise

--- a/gunicorn/workers/sync.py
+++ b/gunicorn/workers/sync.py
@@ -69,8 +69,8 @@ class SyncWorker(base.Worker):
                 # process.
                 continue
 
-            except socket.error as e:
-                if e.args[0] not in (errno.EAGAIN, errno.ECONNABORTED,
+            except EnvironmentError as e:
+                if e.errno not in (errno.EAGAIN, errno.ECONNABORTED,
                         errno.EWOULDBLOCK):
                     raise
 
@@ -95,8 +95,8 @@ class SyncWorker(base.Worker):
                 for listener in ready:
                     try:
                         self.accept(listener)
-                    except socket.error as e:
-                        if e.args[0] not in (errno.EAGAIN, errno.ECONNABORTED,
+                    except EnvironmentError as e:
+                        if e.errno not in (errno.EAGAIN, errno.ECONNABORTED,
                                 errno.EWOULDBLOCK):
                             raise
 
@@ -139,11 +139,11 @@ class SyncWorker(base.Worker):
             else:
                 self.log.debug("Error processing SSL request.")
                 self.handle_error(req, client, addr, e)
-        except socket.error as e:
-            if e.args[0] not in (errno.EPIPE, errno.ECONNRESET):
+        except EnvironmentError as e:
+            if e.errno not in (errno.EPIPE, errno.ECONNRESET):
                 self.log.exception("Socket error processing request.")
             else:
-                if e.args[0] == errno.ECONNRESET:
+                if e.errno == errno.ECONNRESET:
                     self.log.debug("Ignoring connection reset")
                 else:
                     self.log.debug("Ignoring EPIPE")
@@ -181,7 +181,7 @@ class SyncWorker(base.Worker):
             finally:
                 if hasattr(respiter, "close"):
                     respiter.close()
-        except socket.error:
+        except EnvironmentError:
             exc_info = sys.exc_info()
             # pass to next try-except level
             six.reraise(exc_info[0], exc_info[1], exc_info[2])
@@ -193,7 +193,7 @@ class SyncWorker(base.Worker):
                 try:
                     client.shutdown(socket.SHUT_RDWR)
                     client.close()
-                except socket.error:
+                except EnvironmentError:
                     pass
                 raise StopIteration()
             raise

